src/System.Net.Primitives/src/System/Net/IPAddressParser.cs

```diff
@@ -99,33 +184,31 @@ internal static string IPv6AddressToString(ushort[] address, uint scopeId)
                 buffer.Append('%').Append(scopeId);
             }
 
-            return StringBuilderCache.GetStringAndRelease(buffer);
+            return buffer;
         }
 
         private static unsafe void FormatIPv4AddressNumber(int number, char* addressString, ref int offset)
         {
-            int i = offset;
+            offset += number > 99 ? 3 : number > 9 ? 2 : 1;
+
+            int i = 0;
             do
             {
                 int rem;
                 number = Math.DivRem(number, 10, out rem);
-                addressString[--i] = (char)('0' + rem);
+                addressString[offset - ++i] = (char)('0' + rem);
             } while (number != 0);
-            offset = i;
```

Comment:
I see.  Ok.  In that case it'd be worth adding an assert at the beginning validating that number <= 255.  Or better yet, make it a byte rather than an int.